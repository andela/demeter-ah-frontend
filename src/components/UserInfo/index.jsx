import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import './index.scss';
import uploadIcon from '../../assets/images/upload.png';
import { setProfileImage } from '../../store/actions/editProfile';
import FollowBtn from '../FollowButton';

const userInfo = ({ user, setImage, loggedInUser }) => {
  const [userData, setUser] = useState({
    ...user,
    image: user.image || 'https://i.imgur.com/wtjaVfi.png',
  });
  /* istanbul ignore next */
  const loadFilePath = async (file) => {
    const readPath = new FileReader();
    readPath.readAsDataURL(file);
    readPath.onload = e => setUser(prevState => ({ ...prevState, image: e.target.result }));
  };
    /* istanbul ignore next */
  const handleChange = (event) => {
    const [pictureFile] = event.target.files;
    setImage(pictureFile);
    loadFilePath(pictureFile);
  };

  const setUserInfo = () => {
    setUser(prevState => ({
      ...prevState,
      ...user,
      image: user.image || 'https://i.imgur.com/wtjaVfi.png',
    }));
  };

  useEffect(
    () => {
      setUserInfo();
    },
    [user]
  );

  return (
    <div className="userInfo w-full lg:w-10/12 mx-auto flex justify-center lg:justify-start items-center py-10">
      <div className="profileImg relative sm:min-w-32 sm:mr-4 sm:ml-2 md:mr-12">
        <img
          src={userData.image}
          alt=""
          className="prodilePic shadow-md h-32 w-32 md:h-40 md:w-40  lg:h-48 lg:w-48 object-cover rounded-full"
        />
        <div className={`upload absolute rounded-full ${loggedInUser.username !== user.username ? 'hide' : ''}`}>
          <label htmlFor="changePhoto">
            <img
              className=" m-0 h-9 w-9 rounded-full cursor-pointer border-4 border-solid border-white"
              src={uploadIcon}
              alt=""
            />
          </label>
          <input
            id="changePhoto"
            className="hidden"
            type="file"
            onChange={handleChange}
            accept="image/*"
            data-max-size="2000"
          />
        </div>
      </div>
      <div className="info text-left h-full px-2">
        <p className="username sm:max-w-56 md:max-w-xs my-2 font-semibold text-1.5xl md:text-3xl">
          {userData.username}
        </p>
        <p className="fullname mb-2 sm:max-w-56 md:max-w-xs text-sm md:text-1.5xl text-gray-550">{`${userData.firstName}  ${userData.lastName}`}</p>
        <div className="membership flex mb-2">
          <div className="mr-5 flex  sm:flex-row flex-col items-start items-center">
            <strong className="mr-1 text-sm text-left w-full md:text-base font-semibold">
              {userData.followingNo}
            </strong>
            <small className="text-gray-550 text-xs font-semibold">
              Following
            </small>
          </div>
          <div className="flex flex-col sm:flex-row items-start items-center">
            <strong className="mr-1 text-sm text-left w-full md:text-base font-semibold">
              {userData.followersNo}
            </strong>
            <small className="text-gray-550 font-semibold">Followers</small>
          </div>
        </div>
        <p className="bio text-xs sm:max-w-56 md:max-w-86 md:text-sm max-w-86 md:max-w-118 lg:max-w-122">
          {userData.bio === 'null' ? '' : userData.bio}
        </p>
        <FollowBtn viewedUser={user} classes="border border-solid rounded-xxs border-purple-250 w-32 ml-0 text-white text-sm mt-4" />
      </div>
    </div>
  );
};

const mapStateToProps = state => ({
  isCompleted: state.editProfile.isCompleted,
  loggedInUser: state.auth.user,
});

export default connect(mapStateToProps, { setImage: setProfileImage })(
  userInfo
);
