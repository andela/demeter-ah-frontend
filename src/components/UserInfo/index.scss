.userInfo {
  font-family: 'Montserrat', sans-serif;
  .profileImg {
    .upload {
      right: 15%;
      bottom: -5%;
    }

    .hide {
      display: none;
    }
  }

}