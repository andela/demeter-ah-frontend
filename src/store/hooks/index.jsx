import { useEffect } from 'react';
import { getUser } from '../actions/signup';

export const useSetUser = ({ dispatch }) => {
  useEffect(() => {
    dispatch(getUser());
  }, []);
  return [];
};
