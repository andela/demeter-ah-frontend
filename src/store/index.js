import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import combineReducers from './reducers';
import dummyAction from './actions/Signup';

const middlewares = applyMiddleware(thunk);

const store = createStore(
  combineReducers,
  composeWithDevTools(middlewares),
);

store.dispatch(dummyAction);

export default store;
