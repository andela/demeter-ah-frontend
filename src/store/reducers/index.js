import { combineReducers } from 'redux';
import auth from './Auth';
import articleListing from './articleListing';
import resetPassword from './resetPassword';
import changePassword from './changePassword';
import articles from './articles';
import editProfile from './editProfile';
import viewProfile from './viewProfile';
import membership from './Membership';
import viewArticle from './viewArticle';
import relatedArticles from './relatedArticles';
import bookmarks from './bookmarks';
import bookmarkArticle from './bookmarkArticle';
import comments from './Comments';
import search from './search';
import notifications from './notifications';
import rate from './rateArticle';

const reducers = combineReducers({
  auth,
  resetPassword,
  changePassword,
  articles,
  editProfile,
  viewProfile,
  membership,
  articleListing,
  viewArticle,
  relatedArticles,
  bookmarks,
  bookmarkArticle,
  comments,
  search,
  notifications,
  rate
});

export default reducers;
