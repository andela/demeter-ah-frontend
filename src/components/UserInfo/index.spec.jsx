import React from 'react';
import { Provider } from 'react-redux';
import { shallow, mount } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';
import UserInfo from './index';
import store from '../../store';

const user = {
  firstName: 'kevo',
  lastName: 'baba',
  username: 'frankobaba',
  email: 'kevo@me.com',
  image: 'somelink'
};

const setImage = () => ({});

describe('UserInfo Component', () => {
  it('Should render without errors', () => {
    const component = shallow(
      <Router>
        <UserInfo store={store} setImage={setImage} user={user} />
      </Router>
    );
    expect(component).toMatchSnapshot();
  });

  it('Should render along with children component', () => {
    const component = mount(
      <Provider store={store}>
        <Router>
          <UserInfo store={store} setImage={setImage} user={user} />
        </Router>
      </Provider>
    );
    expect(component).toMatchSnapshot();
    expect(component.find('img')).toHaveLength(2);
    expect(component.find('input')).toHaveLength(1);
  });

  it('Should render along with children component wher image is null', () => {
    const component = mount(
      <Provider store={store}>
        <Router>
          <UserInfo store={store} setImage={setImage} user={{...user, image: null }} />
        </Router>
      </Provider>
    );
    expect(component).toMatchSnapshot();
    expect(component.find('img')).toHaveLength(2);
    expect(component.find('input')).toHaveLength(1);
  });
});
