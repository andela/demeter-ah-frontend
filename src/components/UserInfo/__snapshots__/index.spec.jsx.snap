// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInfo Component Should render along with children component 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(userInfo)
        setImage={[Function]}
        store={
          Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
        }
        user={
          Object {
            "email": "kevo@me.com",
            "firstName": "kevo",
            "image": "somelink",
            "lastName": "baba",
            "username": "frankobaba",
          }
        }
      >
        <userInfo
          isCompleted={false}
          loggedInUser={Object {}}
          setImage={[Function]}
          store={
            Object {
              "dispatch": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
              Symbol(observable): [Function],
            }
          }
          user={
            Object {
              "email": "kevo@me.com",
              "firstName": "kevo",
              "image": "somelink",
              "lastName": "baba",
              "username": "frankobaba",
            }
          }
        >
          <div
            className="userInfo w-full lg:w-10/12 mx-auto flex justify-center lg:justify-start items-center py-10"
          >
            <div
              className="profileImg relative sm:min-w-32 sm:mr-4 sm:ml-2 md:mr-12"
            >
              <img
                alt=""
                className="prodilePic shadow-md h-32 w-32 md:h-40 md:w-40  lg:h-48 lg:w-48 object-cover rounded-full"
                src="somelink"
              />
              <div
                className="upload absolute rounded-full hide"
              >
                <label
                  htmlFor="changePhoto"
                >
                  <img
                    alt=""
                    className=" m-0 h-9 w-9 rounded-full cursor-pointer border-4 border-solid border-white"
                    src="test-file-stub"
                  />
                </label>
                <input
                  accept="image/*"
                  className="hidden"
                  data-max-size="2000"
                  id="changePhoto"
                  onChange={[Function]}
                  type="file"
                />
              </div>
            </div>
            <div
              className="info text-left h-full px-2"
            >
              <p
                className="username my-2 font-semibold text-1.5xl md:text-3xl"
              >
                frankobaba
              </p>
              <p
                className="fullname mb-2 text-sm md:text-1.5xl text-gray-550"
              >
                kevo  baba
              </p>
              <div
                className="membership flex mb-2"
              >
                <div
                  className="mr-5 flex  sm:flex-row flex-col items-start items-center"
                >
                  <strong
                    className="md:mr-1 text-sm text-left w-full md:text-base font-semibold"
                  />
                  <small
                    className="text-gray-550 text-xs font-semibold"
                  >
                    Following
                  </small>
                </div>
                <div
                  className="flex flex-col sm:flex-row items-start items-center"
                >
                  <strong
                    className="md:mr-1 text-sm text-left w-full md:text-base font-semibold"
                  />
                  <small
                    className="text-gray-550 font-semibold"
                  >
                    Followers
                  </small>
                </div>
              </div>
              <p
                className="bio text-xs md:text-sm max-w-86 md:max-w-118 lg:max-w-122"
              />
            </div>
          </div>
        </userInfo>
      </Connect(userInfo)>
    </Router>
  </BrowserRouter>
</Provider>
`;

exports[`UserInfo Component Should render along with children component wher image is null 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(userInfo)
        setImage={[Function]}
        store={
          Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
        }
        user={
          Object {
            "email": "kevo@me.com",
            "firstName": "kevo",
            "image": null,
            "lastName": "baba",
            "username": "frankobaba",
          }
        }
      >
        <userInfo
          isCompleted={false}
          loggedInUser={Object {}}
          setImage={[Function]}
          store={
            Object {
              "dispatch": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
              Symbol(observable): [Function],
            }
          }
          user={
            Object {
              "email": "kevo@me.com",
              "firstName": "kevo",
              "image": null,
              "lastName": "baba",
              "username": "frankobaba",
            }
          }
        >
          <div
            className="userInfo w-full lg:w-10/12 mx-auto flex justify-center lg:justify-start items-center py-10"
          >
            <div
              className="profileImg relative sm:min-w-32 sm:mr-4 sm:ml-2 md:mr-12"
            >
              <img
                alt=""
                className="prodilePic shadow-md h-32 w-32 md:h-40 md:w-40  lg:h-48 lg:w-48 object-cover rounded-full"
                src="https://i.imgur.com/wtjaVfi.png"
              />
              <div
                className="upload absolute rounded-full hide"
              >
                <label
                  htmlFor="changePhoto"
                >
                  <img
                    alt=""
                    className=" m-0 h-9 w-9 rounded-full cursor-pointer border-4 border-solid border-white"
                    src="test-file-stub"
                  />
                </label>
                <input
                  accept="image/*"
                  className="hidden"
                  data-max-size="2000"
                  id="changePhoto"
                  onChange={[Function]}
                  type="file"
                />
              </div>
            </div>
            <div
              className="info text-left h-full px-2"
            >
              <p
                className="username my-2 font-semibold text-1.5xl md:text-3xl"
              >
                frankobaba
              </p>
              <p
                className="fullname mb-2 text-sm md:text-1.5xl text-gray-550"
              >
                kevo  baba
              </p>
              <div
                className="membership flex mb-2"
              >
                <div
                  className="mr-5 flex  sm:flex-row flex-col items-start items-center"
                >
                  <strong
                    className="md:mr-1 text-sm text-left w-full md:text-base font-semibold"
                  />
                  <small
                    className="text-gray-550 text-xs font-semibold"
                  >
                    Following
                  </small>
                </div>
                <div
                  className="flex flex-col sm:flex-row items-start items-center"
                >
                  <strong
                    className="md:mr-1 text-sm text-left w-full md:text-base font-semibold"
                  />
                  <small
                    className="text-gray-550 font-semibold"
                  >
                    Followers
                  </small>
                </div>
              </div>
              <p
                className="bio text-xs md:text-sm max-w-86 md:max-w-118 lg:max-w-122"
              />
            </div>
          </div>
        </userInfo>
      </Connect(userInfo)>
    </Router>
  </BrowserRouter>
</Provider>
`;

exports[`UserInfo Component Should render without errors 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Connect(userInfo)
    setImage={[Function]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
    user={
      Object {
        "email": "kevo@me.com",
        "firstName": "kevo",
        "image": "somelink",
        "lastName": "baba",
        "username": "frankobaba",
      }
    }
  />
</Router>
`;
